src/Common/FileCache.h

```diff
@@ -247,6 +256,86 @@ class LRUFileCache final : public IFileCache
     size_t max_stash_element_size;
     size_t enable_cache_hits_threshold;
 
+    struct QueryContext
+    {
+        LRUQueue lru_queue;
+        AccessRecord records;
+
+        size_t cache_size = 0;
+        size_t max_cache_size;
+        size_t ref_count = 0;
+
+        bool skip_download_if_exceeds_query_cache;
+
+        QueryContext(size_t max_cache_size_, bool skip_download_if_exceeds_query_cache_)
+            : max_cache_size(max_cache_size_), skip_download_if_exceeds_query_cache(skip_download_if_exceeds_query_cache_) { }
+
+        void remove(const Key & key, size_t offset, size_t size, std::lock_guard<std::mutex> & cache_lock)
+        {
+            if (!skip_download_if_exceeds_query_cache)
+            {
+                auto record = records.find({key, offset});
+                if (record != records.end())
+                {
+                    lru_queue.remove(record->second, cache_lock);
+                    records.erase({key, offset});
+                }
+            }
+            cache_size -= size;
+        }
+
+        void reserve(const Key & key, size_t offset, size_t size, std::lock_guard<std::mutex> & cache_lock)
+        {
+            if (!skip_download_if_exceeds_query_cache)
+            {
+                auto record = records.find({key, offset});
+                if (record == records.end())
+                {
+                    auto queue_iter = lru_queue.add(key, offset, 0, cache_lock);
+                    record = records.insert({{key, offset}, queue_iter}).first;
+                }
+                record->second->size += size;
+            }
+            cache_size += size;
+        }
+
+        void use(const Key & key, size_t offset, std::lock_guard<std::mutex> & cache_lock)
+        {
+            if (!skip_download_if_exceeds_query_cache)
+            {
+                auto record = records.find({key, offset});
+                if (record != records.end())
+                    lru_queue.moveToEnd(record->second, cache_lock);
+            }
+        }
+
+        size_t getRefCount() const { return ref_count; }
+
+        void incrementRefCount() { ref_count++; }
+
+        void decrementRefCount() { ref_count--; }
+
+        size_t getMaxCacheSize() { return max_cache_size; }
+
+        size_t getCacheSize() { return cache_size; }
+
+        LRUQueue & queue() { return lru_queue; }
+
+        bool isSkipDownloadIfExceed() { return skip_download_if_exceeds_query_cache; }
+    };
+
+    using QueryContextPtr = std::shared_ptr<QueryContext>;
+    using QueryContextMap = std::unordered_map<String, QueryContextPtr>;
```

Comment:
let's make query_map keys as int's instead of strings (same as cache keys), for example by getting hash of the query id string (sipHash)

Reply:
Hello, I want to ask what is the benefit of replacing string with int? Because if we replace string with int, the hash value needs to be calculated once to access the query_map every time, which will cause additional CPU overhead.

Reply:
I am ok to leave it as is with strings. Using int is more efficient in case of many keys, but here it should be a relatively small number.

Reply:
In order to convert query_id from string to int, we need to use hash function. Every find() to query_map needs to convert query_id from string to int through hash function, is there any extra overhead?

Reply:
The hash calculation is relatively not that expensive in the scope where it is called (once on each nextImpl() call). But as I wrote above

> I am ok to leave it as is with strings. Using int is more efficient in case of many keys, but here it should be a relatively small number.

so no conversion is also ok.